File: frontend/lib/api/meta.ts
import { apiGet } from "./client";

export interface MetaStatus {
  data_version: number;
  schemes_count: number;
  amc_count: number;
  nav_count: number;
  file_size_mb: number;
  min_date: string | null;
  max_date: string | null;
  db_path: string;
  is_stale: boolean;
  expected_date: string | null;
  ter_records_count?: number;
  ter_official_schemes?: number;
  flags?: ProductFlags;
}

export interface ProductFlags {
  holdout_portfolios: boolean;
  bl_ui: boolean;
  admin_auth_required: boolean;
  admin_token_configured: boolean;
  amfi_ssl_insecure: boolean;
}

export interface DataQuality {
  schemes_count: number;
  nav_count: number;
  ter_official_schemes: number;
  ter_unknown_schemes: number;
  ter_legacy_schemes: number;
  ter_official_coverage_ratio: number;
  nav_gap_rate: number | null;
  nav_max_date_lag_days?: number | null;
  factor_market_last_nav: string | null;
  factor_momentum_last_nav: string | null;
  summary_table_built_at: string | null;
  ter_official_by_amc: { fund_house: string; official: number; total: number; coverage: number }[];
  ter_official_by_broad_category: { broad_category: string; official: number; total: number; coverage: number }[];
  factor_proxy_codes: Record<string, number>;
}

export interface MetaFilters {
  amcs: string[];
  broad_categories: string[];
  sub_categories: string[];
  options: string[];
  plans: string[];
}

export const getMetaStatus = () => apiGet<MetaStatus>("/api/meta/status");
export const getMetaFilters = (broadCategory?: string) =>
  apiGet<MetaFilters>("/api/meta/filters", { broad_category: broadCategory });
export const getDataQuality = () => apiGet<DataQuality>("/api/meta/data-quality");

File: backend/app/routers/meta.py
from fastapi import APIRouter

from app.core.config import product_flags
from app.db import queries as db
from app.db.connection import get_data_version
from app import amfi_sync
from app.schemas.meta import MetaFilters, MetaStatus

router = APIRouter(prefix="/api/meta", tags=["meta"])


@router.get("/status", response_model=MetaStatus)
def status() -> MetaStatus:
    """One consolidated poll target for the frontend's global date-range store
    (see the migration plan) -- data_version + DB bounds/counts + staleness."""
    stats = db.get_database_stats()
    is_stale, _current_max, expected = amfi_sync.is_database_stale()
    return MetaStatus(
        data_version=get_data_version(),
        schemes_count=stats["schemes_count"],
        amc_count=stats["amc_count"],
        nav_count=stats["nav_count"],
        file_size_mb=stats["file_size_mb"],
        min_date=stats.get("min_date"),
        max_date=stats.get("max_date"),
        db_path=stats["db_path"],
        is_stale=is_stale,
        expected_date=expected,
        ter_records_count=stats.get("ter_count", 0),
        ter_official_schemes=stats.get("ter_official_schemes", 0),
        flags=product_flags(),
    )


@router.get("/data-quality")
def data_quality() -> dict:
    return db.get_data_quality()


@router.get("/filters", response_model=MetaFilters)
def filters(broad_category: str | None = None) -> MetaFilters:
    return MetaFilters(
        amcs=db.get_amcs(),
        broad_categories=db.get_broad_categories(),
        sub_categories=db.get_subcategories(broad_category),
        options=db.get_options_list(),
        plans=db.get_plans_list(),
    )
